libraries/lib-utility/GlobalVariable.h

```diff
@@ -0,0 +1,167 @@
+ @tparam ScopedOnly if true, then enforce RAII for changes of the variable
+ (do not generate Set() or Scope::Commit())
+ */
+template <typename Tag, typename Type, Type (*initializer)() = nullptr,
+   bool ScopedOnly = true>
+class GlobalVariable {
+public:
+   using variable_type = GlobalVariable;
+   using stored_type = Type;
+   using mutable_type = std::remove_const_t<Type>;
+
+   //! Get the installed value
+   static stored_type& Get()
+   {
+      return Instance();
+   }
+
+   //! Move in a new value, move out and return the previous
+   /*! Won't compile unless ScopedOnly is false */
+   [[nodiscard]] static mutable_type Set(mutable_type replacement)
+   {
+      return Assign<!ScopedOnly>(std::move(replacement));
+   }
+
+   //! RAII guard for temporary installation of a value; movable
+   /*!
+    Constructor captures previous value, installs a new value;
+    destructor restores the previous value, unless moved-from or `Commit`ted
+
+    Not required to have stack-like lifetime.  Non-nested lifetimes for
+    distinct Scope objects is not prevented and may have surprising results.
+    */
+   class Scope
+   {
+   public:
+      explicit Scope(mutable_type value)
+         : m_previous{ Assign(std::move(value)) }
+      {}
+      Scope(Scope &&other) = default;
+      Scope &operator=(Scope &&other) = default;
+      ~Scope()
+      {
+         if constexpr (ScopedOnly)
+            Assign(std::move(m_previous));
+         else if (m_previous)
+            Assign(std::move(*m_previous));
+      }
+      /*! Won't compile unless ScopedOnly is false */
+      void Commit()
+      {
+         static_assert(!ScopedOnly);
+         m_previous.reset();
+      }
+      bool HasValue() const
+      {
+         if constexpr (ScopedOnly)
+            return true;
+         else
+            return m_previous.has_value();
+      }
+
+   private:
+      std::conditional_t<ScopedOnly, mutable_type, std::optional<mutable_type>>
+         m_previous;
+   };
+
+   /*! @brief Can guarantee that the global variable's lifetime encloses
+    those of other objects of static duration
+    
+    This is like the "nifty counter" idiom.  The guarantee is not automatic,
+    but you can define a `static` instance in a multiply included header file,
+    which creates multiple Initializers, but only one of them causes the
+    initial construction of the variable.
+
+    This makes it safe for other calls to Get() to occur inside the bodies
+    of both the constructor and destructor of another long-lived object.  Such
+    objects can be freely defined by other translation units that include the
+    header with no extra effort.
+    */
+   struct Initializer { Initializer() { Instance(); } };
+
+private:
+   //! Use static functions only.  Don't directly construct this.
+   GlobalVariable() = delete;
+
+   //! Generate the static variable
+#ifdef _WIN32
+   __declspec(dllexport)
+#endif
+   static mutable_type &Instance()
```

Comment:
The name `Instance` looks a bit misleading to me, considering that return type isn't a `GlobalVariable<>`

Reply:
But this is a private member function.  Users of the class use it only indirectly by calling `Get()` or using `Initializer`.